aten/src/ATen/cuda/CUDABlas.cpp

```diff
@@ -0,0 +1,325 @@
+  // and at least as big the result requires (even if the value won't
+  // be used).
+  if (n <= 1)
+    *ldc = std::max<int64_t>(m, 1);
+
+  if (transa_) {
+    if (m <= 1)
+      *lda = std::max<int64_t>(k, 1);
+  } else {
+    if (k <= 1)
+      *lda = std::max<int64_t>(m, 1);
+  }
+
+  if (transb_) {
+    if (k <= 1)
+      *ldb = std::max<int64_t>(n, 1);
+  } else {
+    if (n <= 1)
+      *ldb = std::max<int64_t>(k, 1);
+  }
+}
+} // anonymous namespace
+
+namespace at {
+namespace cuda {
+namespace blas {
+
+/* LEVEL 3 BLAS FUNCTIONS */
+
+#define GEMM_CHECK_ARGVALUES(Dtype)           \
+  do {                                        \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, m); \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, n); \
+    CUDABLAS_NONNEGINT_CHECK(gemm<Dtype>, k); \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, lda);  \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, ldb);  \
+    CUDABLAS_POSINT_CHECK(gemm<Dtype>, ldc);  \
+  } while (0)
+
+template <>
+void gemm<double>(CUDABLAS_GEMM_ARGTYPES(double)) {
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  cublasOperation_t opa = _cublasOpFromChar(transa);
+  cublasOperation_t opb = _cublasOpFromChar(transb);
+  _cublasAdjustLdLevel3(transa, transb, m, n, k, &lda, &ldb, &ldc);
+  GEMM_CHECK_ARGVALUES(double);
+  TORCH_CUDABLAS_CHECK(cublasSetStream(handle, stream));
+  TORCH_CUDABLAS_CHECK(cublasDgemm(
+      handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc));
+}
+
+template <>
+void gemm<float>(CUDABLAS_GEMM_ARGTYPES(float)) {
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  cublasOperation_t opa = _cublasOpFromChar(transa);
+  cublasOperation_t opb = _cublasOpFromChar(transb);
+  _cublasAdjustLdLevel3(transa, transb, m, n, k, &lda, &ldb, &ldc);
+  GEMM_CHECK_ARGVALUES(float);
+  TORCH_CUDABLAS_CHECK(cublasSetStream(handle, stream));
+  TORCH_CUDABLAS_CHECK(cublasSgemm(
+      handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc));
+}
+
+template <>
+void gemm<at::Half>(CUDABLAS_GEMM_ARGTYPES(at::Half)) {
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  cublasOperation_t opa = _cublasOpFromChar(transa);
+  cublasOperation_t opb = _cublasOpFromChar(transb);
+  float falpha = alpha;
+  float fbeta = beta;
+  _cublasAdjustLdLevel3(transa, transb, m, n, k, &lda, &ldb, &ldc);
+  GEMM_CHECK_ARGVALUES(at::Half);
+  TORCH_CUDABLAS_CHECK(cublasSetStream(handle, stream));
+#ifdef __HIP_PLATFORM_HCC__
+  TORCH_CUDABLAS_CHECK(rocblas_gemm_ex(
+      handle,
+      opa,
+      opb,
+      m,
+      n,
+      k,
+      &falpha,
+      a,
+      rocblas_datatype_f16_r,
+      lda,
+      b,
+      rocblas_datatype_f16_r,
+      ldb,
+      &fbeta,
+      c,
+      rocblas_datatype_f16_r,
+      ldc,
+      c,
+      rocblas_datatype_f16_r,
+      ldc,
+      rocblas_datatype_f32_r,
+      rocblas_gemm_algo_standard,
+      0,
+      0,
+      NULL,
+      NULL));
+#else
+
+#if CUDA_VERSION >= 9000
```

Comment:
nit, I like indenting nested preprocessor macros like `#  if CUDA_VERSION >= 9000`, I find it makes following the indentation structure easier.

Reply:
@syed-ahmed Hey, apparently we missed this ifdef, don't we not support CUDA 8 any more?

Reply:
good nit, applied